When polling multiple sites, if a controller drops, no data is scraped.

Hello, 

I am trying to use unpoller to poll multiple controllers simultaneously. I keep coming across an event where when one of the controllers becomes unavailable for polling, usually through a timeout or disconnected network, unpoller then tries to reconnect to that site. During these times of reconnection no information from any online and connected site is polled. I am seeking a way to correct this behavior in a way where I can still get polled data while unpoller retries connecting to the down/unavailable controller. 

This has occurred across several deployments and I cannot seem to find a way to correct it. 

I am running unpoller in docker within a bridged network that also contains Grafana, influxdb 1.8, grafana renderer, and a pdf report generator. The end goal is to create a monitoring system for remote sites I manage which feedback to a comprehensive controller dashboard and a single overview panel that displays all the controller's mission critical data. 

On Docker Desktop (win 10 WSL)
  I have done both env and docker compose configurations, both of which can see both controllers simultaneously, but will fail if a controller goes down. 

Synology
  I was asked to migrate the application stack to a dedicated synology. I've been using the env variables in the advanced options without any troubles, but once again, if one site cannot be seen, no data is reported. 

Test site, polling correctly. 

`2022/09/08 20:40:22.599671 updateweb.go:213: [DEBUG] Requesting https://192.168.96.
142:443/proxy/network/api/stat/sites, with params: false, cookies: 1               
2022/09/08 20:40:22.672652 updateweb.go:213: [DEBUG] Requested https://192.168.96.1
42:443/proxy/network/api/stat/sites: elapsed 73ms, returned 700 bytes              
2022/09/08 20:40:22.672964 updateweb.go:213: [DEBUG] Found 1 site(s): default      
2022/09/08 20:40:22.672982 updateweb.go:213: [DEBUG] Polling Controller for IDS Eve
nts, site Default (default)                                                        
2022/09/08 20:40:22.673067 updateweb.go:213: [DEBUG] Requesting https://192.168.96.
142:443/proxy/network/api/s/default/stat/ips/event, with params: true, cookies: 1  
2022/09/08 20:40:22.703515 updateweb.go:213: [DEBUG] Requested https://192.168.96.1
42:443/proxy/network/api/s/default/stat/ips/event: elapsed 30ms, returned 40 bytes 
2022/09/08 20:40:22.703578 updateweb.go:213: [DEBUG] Polling Controller for Alarms,
 site Default (default)                                                            
2022/09/08 20:40:22.703642 updateweb.go:213: [DEBUG] Requesting https://192.168.96.
142:443/proxy/network/api/s/default/list/alarm, with params: false, cookies: 1     
2022/09/08 20:40:22.768773 updateweb.go:213: [DEBUG] Requested https://192.168.96.1
42:443/proxy/network/api/s/default/list/alarm: elapsed 65ms, returned 737 bytes    
2022/09/08 20:40:22.817117 logger.go:17: [INFO] UniFi Metrics Recorded. Site: 1, Cl
ient: 0, Gateways: 0, UAP: 1, USW: 0, IDS/Event/Alarm/Anomaly: 0/0/0/0, DPI Site/Cl
ient: 1/0, Point: 12, Fields: 401, Err: 0, Dur: 48ms`

Same test site, with additional intentionally unreachable controller. 

`2022/09/08 20:49:58.456325 updateweb.go:213: [DEBUG] Requesting https://192.168.
96.142:443/proxy/network/api/stat/sites, with params: false, cookies: 1         
2022/09/08 20:49:58.525000 updateweb.go:213: [DEBUG] Requested https://192.168.9
6.142:443/proxy/network/api/stat/sites: elapsed 69ms, returned 700 bytes        
2022/09/08 20:49:58.525166 updateweb.go:213: [DEBUG] Found 1 site(s): default   
2022/09/08 20:49:58.525200 updateweb.go:213: [DEBUG] Polling Controller, retreiv
ing Site DPI data, site Default (default)                                       
2022/09/08 20:49:58.525267 updateweb.go:213: [DEBUG] Requesting https://192.168.
96.142:443/proxy/network/api/s/default/stat/sitedpi, with params: true, cookies:
 1                                                                              
2022/09/08 20:49:58.562846 updateweb.go:213: [DEBUG] Requested https://192.168.9
6.142:443/proxy/network/api/s/default/stat/sitedpi: elapsed 38ms, returned 60 by
tes                                                                             
2022/09/08 20:49:58.562967 updateweb.go:213: [DEBUG] Polling Controller, retreiv
ing Client DPI data, site Default (default)                                     
2022/09/08 20:49:58.563015 updateweb.go:213: [DEBUG] Requesting https://192.168.
96.142:443/proxy/network/api/s/default/stat/stadpi, with params: true, cookies: 
1
2022/09/08 20:48:28.595495 updateweb.go:213: [DEBUG] Requested https://192.168.9
6.142:443/proxy/network/api/s/default/stat/stadpi: elapsed 23ms, returned 30 byt
es                                                                              
2022/09/08 20:48:28.595549 updateweb.go:213: [DEBUG] Polling Controller, retreiv
ing UniFi Clients, site Default (default)                                       
2022/09/08 20:48:28.595619 updateweb.go:213: [DEBUG] Requesting https://192.168.
96.142:443/proxy/network/api/s/default/stat/sta, with params: false, cookies: 1 
2022/09/08 20:48:28.614301 updateweb.go:213: [DEBUG] Requested https://192.168.9
6.142:443/proxy/network/api/s/default/stat/sta: elapsed 19ms, returned 30 bytes 
2022/09/08 20:48:28.614410 updateweb.go:213: [DEBUG] Requesting https://192.168.
96.142:443/proxy/network/api/s/default/stat/device, with params: false, cookies:
 1                                                                              
2022/09/08 20:48:28.641092 updateweb.go:213: [DEBUG] Requested https://192.168.9
6.142:443/proxy/network/api/s/default/stat/device: elapsed 27ms, returned 20147 
bytes                                                                           
2022/09/08 20:48:28.642545 updateweb.go:213: [DEBUG] Unmarshalling Device Type: 
uap, site Default (default)                                                     
2022/09/08 20:48:28.644328 updateweb.go:193: [INFO] Re-authenticating to UniFi C
ontroller: https://192.168.96.231:443                                           
2022/09/08 20:48:28.644352 updateweb.go:213: [DEBUG] Requesting https://192.168.
96.231:443/ to determine API paths                                              
2022/09/08 20:48:31.735357 logger.go:27: [ERROR] metric fetch for InfluxDB faile
d: re-authenticating to https://192.168.96.231:443: unifi controller: making req
uest: Get "https://192.168.96.231:443/": dial tcp 192.168.96.231:443: connect: n
o route to host  
`

I can tell its seeing the connected site still. The default name is correct as is the latency and size. Yet if I go to a Time Series panel in Grafana and view it in table mode, I can see no data is making its way through. Once I correct the 'down' site, data polling starts coming back. 

My biggest concern is that since I'm using Influxdb to scrape the data, any time where I cannnot collect data is equivalent to being unable to monitor those remote controllers and having no record of their activity during that time. 

I am using ENV's as laid out on this page of the documentation site: https://unpoller.com/docs/install/configuration
  (controller_0_  and controller_1_   are being used respectively)

Any help or suggestions would be appreciated, thanks!

--- a/pkg/inputunifi/interface.go
+++ b/pkg/inputunifi/interface.go
@@ -200,6 +200,8 @@ func (u *InputUnifi) Events(filter *poller.Filter) (*poller.Events, error) {
 		filter = &poller.Filter{}
 	}
 
+	var collectionErrors []error
+
 	for _, c := range u.Controllers {
 		if filter.Path != "" && !strings.EqualFold(c.URL, filter.Path) {
 			continue
@@ -207,10 +209,19 @@ func (u *InputUnifi) Events(filter *poller.Filter) (*poller.Events, error) {
 
 		events, err := u.collectControllerEvents(c)
 		if err != nil {
-			return nil, err
+			// Log error but continue to next controller
+			u.LogErrorf("Failed to collect events from controller %s: %v", c.URL, err)
+			collectionErrors = append(collectionErrors, fmt.Errorf("%s: %w", c.URL, err))
+			continue
 		}
 
 		logs = append(logs, events...)
+	}
+
+	// Return collected events even if some controllers failed
+	// Only return error if ALL controllers failed and we have no events
+	if len(logs) == 0 && len(collectionErrors) > 0 {
+		return nil, collectionErrors[0]
 	}
 
 	return &poller.Events{Logs: logs}, nil
@@ -229,6 +240,8 @@ func (u *InputUnifi) Metrics(filter *poller.Filter) (*poller.Metrics, error) {
 		filter = &poller.Filter{}
 	}
 
+	var collectionErrors []error
+
 	// Check if the request is for an existing, configured controller (or all controllers)
 	for _, c := range u.Controllers {
 		if filter.Path != "" && !strings.EqualFold(c.URL, filter.Path) {
@@ -238,10 +251,23 @@ func (u *InputUnifi) Metrics(filter *poller.Filter) (*poller.Metrics, error) {
 
 		m, err := u.collectController(c)
 		if err != nil {
-			return metrics, err
+			// Log error but continue to next controller
+			u.LogErrorf("Failed to collect metrics from controller %s: %v", c.URL, err)
+			collectionErrors = append(collectionErrors, fmt.Errorf("%s: %w", c.URL, err))
+			continue
 		}
 
 		metrics = poller.AppendMetrics(metrics, m)
+	}
+
+	// If we collected data from at least one controller, return success
+	if len(metrics.Devices) > 0 || len(metrics.Clients) > 0 {
+		return metrics, nil
+	}
+
+	// If all controllers failed and we had errors, return the first error
+	if len(collectionErrors) > 0 {
+		return metrics, collectionErrors[0]
 	}
 
 	if filter.Path == "" || len(metrics.Clients) != 0 {

